### DISCUSSION TITLE
Associate information with SelectedField

### DISCUSSION THREAD

**Discussion by nk-coding** [graphql-java/graphql-java#2653] [Q&A] [answered]
Hi,
I have a use-case, where I need to do some analysis of the whole query when I first encounter it.
I currently do this by recursively traversing the selectionSet on the `DataFetchingEnvironment`, and generating some information for each `SelectedField`.  
However, I need to access this information later in the `DataFetcher` which actually resolves this field.
What I would like to do is associate some information with a `SelectedField`.  
Is something like that possible? If yes, how could I achieve this?  

Greetings,
Niklas

**Comment by bbakerman** [comment, marked as answer]
SelectedField itself has not extension mechanism to allow you to assocuate data with it per se.

However there are 2 "context" objects in graphql-java that you could use to hold a `Map<SelectedField,YourData>` that allows you to have access to this data

See https://www.graphql-java.com/blog/deep-dive-data-fetcher-results/

Local context is what I would recommend - it passes down from parent field to child ones (including grandchildren unless changed).

```

        Map<SelectedField,YourData> fieldMap = buildThis(env);
        GraphQLContext context = GraphQLContext.newContext().of("fieldMap", fieldMap);
        return DataFetcherResult.newResult().data(fieldData).localContext(context).build();
```

There is also a global GraphQLContext that spans ALL field fetches for that request - that could also work but I think local context is safer and more targeted,

**Reply by nk-coding** [reply]
First of all, thanks for the answer.  
Sadly, this only partially solves my problem. I thought of putting a map with the SelectedField as key into one of the context maps, however in the sub-DataFetchers, I could only access the Field (or MergedField) via the DataFetchingEnvironment, and therefore could not retrieve the correct key for the map. Is there a way to find out which SelectedField is currently fetched?

**Reply by bbakerman** [reply]
>  I thought of putting a map with the SelectedField as key into one of the context maps, however in the sub-DataFetchers, I could only access the Field (or MergedField) via the DataFetchingEnvironment, and therefore could not retrieve the correct key for the map. 

yeah your right - the SelectedField object is created per data fetcher (albeit under the covers there is a canonical object but I digress)

So you can use it as a key

I think you can however use "field path".  

```
query {
   x {
      y {
         z
   }
```

leads to a field path of `["x","y","z"]` or "x/y/z" if you use `graphql.execution.ResultPath` - 

You can use `graphql.execution.ExecutionStepInfo#getPath` to get the path to a specific field via `graphql.schema.DataFetchingEnvironment#getExecutionStepInfo` - so that is a much better key to represent each field

Remember how that ResultPaths can be for list items like `x/y[23]/z` so `graphql.execution.ResultPath#getKeysOnly` might be best for this

**Reply by nk-coding** [reply]
> So you can use it as a key

I think that should mean I *can't* use it as a key? Otherwise, it would be good to know how, because I could not find a way

> I think you can however use "field path".

That could be the solution for me, however I encountered several problems:
1. How to get the ResultPath for a SelectedField? I could build it manually, but there might be a better way to do this
2. (More important)
I have a query which looks like this:
```gql
query {
  node(id: "test") {
    ...on ARoot {
      leafs(orderBy: { field: TEXT }) {
        nodes {
           id
        }
      }
    }
    ...on BRoot {
      leafs(orderBy: { field: ID }) {
        nodes {
           id
        }
      }
    }
  }
}
```
I have to differentiate between the two leafs fields (because they are represented by different SelectedFields and I need to associate different data). However, in the "field path", both are represented by `/node/leafs`.  
I think that I could build my own tree, where a field is identified by name (name or alias), and parent type (I can get this from a DataFetchingEnvironment by getParent()). But I'm not sure, if this approach covers all corner cases, and I also hope that there is a better approach

**Reply by nk-coding** [reply]
> I think that I could build my own tree, where a field is identified by name (name or alias), and parent type

I implemented this approach and it seems to work, therefore I mark this thread as answer  
However, if anyone knows a better appraoch, please let me know
